feat: centralize config with Pydantic Settings and .env support

jaisw7 opened the pull request:
## Summary

Centralize all dnet configuration into a unified Pydantic Settings module with `.env` file support and auto-generation of [.env.example](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/.env.example:0:0-0:0).

## Motivation

Configuration was scattered across multiple files with inconsistent patterns:
- [utils/logger.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/utils/logger.py:0:0-0:0) used `dotenv` + raw `os.getenv()`
- [utils/grpc_config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/utils/grpc_config.py:0:0-0:0) used raw `os.getenv()`
- [core/observability.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/core/observability.py:0:0-0:0) used raw `os.getenv()`
- [utils/repack.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/utils/repack.py:0:0-0:0) used raw `os.getenv()`
- [api/config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/api/config.py:0:0-0:0) and [shard/config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/shard/config.py:0:0-0:0) used dataclasses

This made it difficult to know what environment variables were available and led to duplication. The new approach provides:
- Single source of truth for all configuration
- Type validation via Pydantic
- Automatic `.env` file loading
- Self-documenting [.env.example](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/.env.example:0:0-0:0) that stays in sync with code

## Changes

- Add `pydantic-settings>=2.0.0` to dependencies, remove `python-dotenv` from dev deps
- Create centralized [src/dnet/config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/config.py:0:0-0:0) with all settings classes
- Refactor scattered `os.getenv` calls to use centralized settings with fallbacks
- Deprecate [api/config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/api/config.py:0:0-0:0) and [shard/config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/shard/config.py:0:0-0:0) with backward compatibility layer
- Update CLI entry points to use settings for default argument values
- Add [scripts/generate_env_example.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/scripts/generate_env_example.py:0:0-0:0) to auto-generate [.env.example](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/.env.example:0:0-0:0)
- Add pre-commit hook to regenerate [.env.example](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/.env.example:0:0-0:0) when [config.py](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/src/dnet/config.py:0:0-0:0) changes
- Update [Makefile](cci:7://file:///home/jaiswal0/Desktop/dria/repo/dnet/Makefile:0:0-0:0) with `env-example` target, remove `--env-file .env`

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactoring
- [ ] Performance improvement
- [ ] CI/CD improvement

## Testing

- [x] Tests pass locally
- [ ] Added new tests for the changes
- [ ] Existing tests updated

**Manual verification:**
```bash
# Test settings load correctly
uv run python -c "from dnet.config import get_settings; s = get_settings(); print(s.api.http_port)"
# Output: 8080

# Test env var overrides
DNET_API_HTTP_PORT=9999 uv run python -c "from dnet.config import get_settings; print(get_settings().api.http_port)"
# Output: 9999

# Test .env.example generation
make env-example

jaisw7 committed:
feat: centralize config with Pydantic Settings and .env support

- Add pydantic-settings dependency, remove python-dotenv
- Create centralized config module (src/dnet/config.py) with all settings
- Refactor scattered os.getenv calls to use centralized settings:
  - utils/logger.py
  - utils/grpc_config.py
  - core/observability.py
  - utils/repack.py
- Deprecate api/config.py and shard/config.py with backward compat
- Update CLI entry points to use settings for default values
- Add script to auto-generate .env.example from settings
- Add pre-commit hook to regenerate .env.example when config.py changes
- Update Makefile with env-example target

jaisw7 committed:
refactor: update shard/runtime.py and policies to use centralized settings

- Replace deprecated config imports with centralized dnet.config
- Add RuntimeKVCacheConfig for mutable per-request KV cache updates
- Update plan_policy to accept TopologySettings

jaisw7 committed:
fix(tests): use explicit base_dir in repack tests

Use explicit base_dir parameter instead of relying on DNET_REPACK_DIR
env var which doesn't work reliably with cached settings.

jaisw7 committed:
fix: update tests and adapters for new settings API

- Update test_shard_runtime.py to use kv_cache_config instead of compute_config.kv_cache
- Update test_wire_dtype_bf16_mapping to use env var approach
- Update adapters/ring.py to use TransportSettings from centralized config
- Fix _get_repack_base_dir to check env var first (enables tests to monkeypatch)
- Update test_utils_repack_ops.py with explicit base_dir and working env var approach

jaisw7 committed:
fix: update test_ring_adapter.py to use TransportSettings

jaisw7 committed:
chore: bump dnet patch version

jaisw7 committed:
ci: add concurrency to cancel in-progress runs on new pushes

erhant reviewed (approved):
LGTM & good one!
Just 2 very small comments added.

jaisw7 committed:
fix: add prefetch_mode to ComputeSettings and use mutable RuntimeComputeConfig

- Add prefetch_mode field to ComputeSettings in config.py
- Add RuntimeComputeConfig mutable class in runtime.py
- Create _compute_config mutable instance in ShardRuntime.__init__
- Update compute_config property to return mutable config for policy mutations

Fixes: 'ComputeSettings' object has no attribute 'prefetch_mode'

jaisw7 commented on .env.example:
```diff
@@ -1,5 +1,97 @@
+# Repacked layers directory
+DNET_REPACK_DIR=~/.dria/dnet/repacked_layers
+# Log files directory
+DNET_LOG_DIR=~/.dria/dnet
```
Sounds good! Addressed in the revision

jaisw7 committed:
chore: improve storage paths defaults

- Change log_dir default to ~/.dria/dnet/logs
- Use Path.home() instead of tilde in repack.py for cross-platform compat

jaisw7 committed:
fix: allow servers to initialize for 10sec

andthattoo committed:
Merge branch 'master' into centralize_config
